v3_async_wip/v3_async_wip/mqtt_client.py

```diff
@@ -329,73 +472,123 @@ async def connect(self):
+                        )
+                        self._reconnect_daemon.cancel()
+                        self._reconnect_daemon = None
+
+                        # NOTE: Because a connection could still complete after cancellation due to
+                        # it being in flight, this means that it's possible a connection could be
+                        # established without a running reconnect daemon, even if auto_reconnect
+                        # is enabled. This could be remedied fairly easily if so desired, but I've
+                        # chosen to leave it out for simplicity.
+                    else:
+                        logger.debug(
+                            "Reconnect daemon was started on a previous connect. Leaving it alone."
+                        )
+                    raise
+                finally:
+                    # Pending operation is completed
+                    self._pending_connect = None
 
             else:
                 logger.debug("Already connected!")
 
+    async def _do_connect(self):
+        """Connect, start network loop, and wait for response"""
+
+        # Paho Connect
+        logger.debug("Attempting connect using port {}...".format(self._port))
+        try:
+            rc = await self._loop.run_in_executor(
+                None,
+                functools.partial(
+                    self._mqtt_client.connect,
+                    host=self._hostname,
+                    port=self._port,
+                    keepalive=self._keep_alive,
+                ),
+            )
+            rc_msg = mqtt.error_string(rc)
+            logger.debug("Connect returned rc {} - {}".format(rc, rc_msg))
+        # TODO: more specialization of errors to indicate which are/aren't retryable
+        except asyncio.CancelledError:
+            # Handled in outer method
+            raise
+        except Exception as e:
+            raise MQTTConnectionFailedError(message="Failure in Paho .connect()") from e
+
+        if rc != mqtt.MQTT_ERR_SUCCESS:
+            # NOTE: This block should probably never execute. Paho's .connect() is
+            # supposed to only return success or raise an exception.
+            logger.warning("Unexpected rc {} from Paho .connect()".format(rc))
+            # MQTTConnectionFailedError expects a connack rc, but this is a regular rc.
+            # So chain a regular mqtt exception into a connection mqtt exception.
+            try:
+                raise MQTTError(rc=rc)
+            except MQTTError as e:
+                raise MQTTConnectionFailedError(message="Unexpected Paho .connect() rc") from e
+
+        # Start Paho network loop, and store the task. This task will complete upon disconnect
+        # whether due to invocation of .disconnect(), or an unexpected network drop.
+        #
+        # NOTE: If the connect attempt is cancelled, the network loop cannot be stopped.
```

Comment:
>NOTE: If the connect attempt is cancelled, the network loop cannot be stopped. [](http://example.com/codeflow?start=10&length=78)

"cannot be stopped" needs more detail.  Why can't it be stopped? Paho won't let it stop? We don't provide an API? We don't have a good place/time to stop it?  Some other reason?

Reply:
It's because when using `loop_forever()`, Paho is in charge of the loop's lifecycle. The loop will exit when a connection is lost, whether due to manual disconnect, drop, or even connection failure. There is no way to manually stop it.

I will add this detail to the inline implementation note.